### DISCUSSION TITLE
secp: Why the additional level of abstraction in keys?

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1378] [Q&A] [answered]
In `rust-secp256k1` we have `secp256k1::PublicKey` and `ffi::PublicKey` but for secret keys we only have `secp256k1::SecretKey`, I've been trying to figure out what the additional level of abstraction gives us and why we have it for pubkeys but not secret keys? None of my ideas so far is fully satisfying 

- Its something to do with `repr(C)`?
- It would allow us to only call safe code in `secp256k1` (we are not currently doing this)
- It would enable us to fully conceal all interaction with `libsecp256k1` within `secp256k1-sys` (we are not currently doing this, not sure if this is a benefit or not)

Can someone enlighten me please?

**Comment by tcharding** [comment]
If we fully encapsulated the FFI layer would this allow us to stabalize `rust-secp256k1` and leave `secp256k1-sys` eternally unstable?

**Comment by apoelstra** [comment, marked as answer]
`ffi::PublicKey` directly maps to the `secp256k1_public_key` type in the C code. It exposes its underlying bytestring, again to match the C code, although users are not allowed to depend in any way on its contents (just like in the C code). Meanwhile `secp256k1::PublicKey` is a "safe" wrapper around this which basically just avoids exposing the underlying bytestring.

There is no `ffi::SecretKey` because there is no corresponding type in the C code. The C code just uses `unsigned char*` for secrets.


> * Its something to do with repr(C)?

No, though it is nice not to have `repr(C)` on types exported from the main library.

> * It would allow us to only call safe code in secp256k1 (we are not currently doing this)

I don't understand this. We definitely need to call FFI functions from secp256k1 no matter what we do with our pubkey type.

> * It would enable us to fully conceal all interaction with libsecp256k1 within secp256k1-sys (we are not currently doing this, not sure if this is a benefit or not)

This isn't a benefit. The whole point of secp-sys existing is that there's a way for users outside this project to interact with the libsecp bindings.

**Reply by tcharding** [reply]
Thanks for the explanation man.

> `ffi::PublicKey` directly maps to the `secp256k1_public_key` type in the C code. It exposes its underlying bytestring, again to match the C code, although users are not allowed to depend in any way on its contents (just like in the C code). Meanwhile `secp256k1::PublicKey` is a "safe" wrapper around this which basically just avoids exposing the underlying bytestring.

Cool, I had this bit right in my head.

> There is no `ffi::SecretKey` because there is no corresponding type in the C code. The C code just uses `unsigned char*` for secrets.

I missed this piece of the puzzle!

> > * Its something to do with repr(C)?
> 
> No, though it is nice not to have `repr(C)` on types exported from the main library.
> 
> > * It would allow us to only call safe code in secp256k1 (we are not currently doing this)
> 
> I don't understand this. We definitely need to call FFI functions from secp256k1 no matter what we do with our pubkey type.
> 
> > * It would enable us to fully conceal all interaction with libsecp256k1 within secp256k1-sys (we are not currently doing this, not sure if this is a benefit or not)
> 
> This isn't a benefit. The whole point of secp-sys existing is that there's a way for users outside this project to interact with the libsecp bindings.

Coupled with the comment below by @Kixunil I get it now. Cheers.

**Comment by Kixunil** [comment]
The library is basically following the recommendation of Rust team on how to make bindings: have a `-sys` library that only contains `extern fn` and C-equivalent types and then have a thin safe layer on top of that ensuring correct usage of the library. The main benefit IMO is clarity: you can easily check that the bindings match (there's even a tool to auto-generate them) and then the compiler helps you do the rest. There's also theoretical niche benefit that if you're dissatisfied with the abstraction you can swap it. (Note that each external library may link to Rust project only once.)

**Reply by tcharding** [reply]
ooooh, I did not know this. Do you mind if I paraphrase this and add it to the `secp256k1-sys` module as docs someplace?

**Reply by Kixunil** [reply]
Seems a bit too much, people usually know about the recommendation. That being said I couldn't find where I've seen the recommendation so if we find it we could have a short note pointing to it.

**Reply by apoelstra** [reply]
Heh, yeah, it looks like the -sys split happened in https://github.com/rust-bitcoin/rust-secp256k1/pull/169 and nobody was questioning why we were doing it or anything. I think we sorta cargo-culted it from openssl-sys and other such crates.

**Reply by sanket1729** [reply]
IMO the main benefit of `-sys` crate is that other implementers are free to write their own high-level level APIs using the `-sys` crate. A downstream code can then potentially depend on them both.

**Reply by Kixunil** [reply]
Yep, because Rust can link to only one version of a library (but we use the symbol renaming hack for vendored versions which makes it less of an issue).
